google-cloud-storage/src/test/java/com/google/cloud/storage/ByteRangeSpecTest.java

```diff
@@ -441,129 +444,168 @@ private static void threeWayEqual(
-                  rangeClosed(1L, effectiveMax - 1)),
-              RangeScenario.of(ByteRangeSpec.explicitClosed(1L, null), 1L, INF, INF, rangeOpen(1)),
-              RangeScenario.of(
-                  ByteRangeSpec.explicitClosed(null, effectiveMax),
-                  0,
-                  effectiveMax,
-                  effectiveMax,
-                  rangeClosed(0, effectiveMax)),
-              RangeScenario.of(
-                  ByteRangeSpec.explicitClosed(INF, null), INF, INF, INF, rangeOpen(INF)),
-              RangeScenario.of(
-                  ByteRangeSpec.explicitClosed(1L, effectiveMax),
-                  1L,
-                  effectiveMax,
-                  effectiveMax - 1,
-                  rangeClosed(1L, effectiveMax)));
+          Streams.concat(
+              // expect that, if the range is effectively LeftClosed only, all upper bounds should
+              // be EFFECTIVE_INFINITY and the requests should be open-ended
+              RangeScenario.expectThat()
+                  .beginOffset(1L)
+                  .endOffset(EFFECTIVE_INFINITY)
+                  .endOffsetInclusive(EFFECTIVE_INFINITY)
+                  .length(EFFECTIVE_INFINITY)
+                  .httpRange(headerRangeOpen(1L))
+                  .isApplicableTo(
+                      ByteRangeSpec.relativeLength(1L, null),
+                      ByteRangeSpec.explicit(1L, null),
+                      ByteRangeSpec.explicitClosed(1L, null)),
+              // expect that, we can start reading from effective_infinity with not upper bound
+              RangeScenario.expectThat()
+                  .beginOffset(EFFECTIVE_INFINITY)
+                  .endOffset(EFFECTIVE_INFINITY)
+                  .endOffsetInclusive(EFFECTIVE_INFINITY)
+                  .length(EFFECTIVE_INFINITY)
+                  .httpRange(headerRangeOpen(EFFECTIVE_INFINITY))
+                  .isApplicableTo(
+                      ByteRangeSpec.relativeLength(EFFECTIVE_INFINITY, null),
+                      ByteRangeSpec.explicit(EFFECTIVE_INFINITY, null),
+                      ByteRangeSpec.explicitClosed(EFFECTIVE_INFINITY, null)),
+              // expect that, we can read up to Long.MAX_VALUE - 1
+              RangeScenario.expectThat()
+                  .beginOffset(0L)
+                  .endOffset(effectiveMax_0)
+                  .endOffsetInclusive(effectiveMax_1)
+                  .length(effectiveMax_0)
+                  .bounded()
+                  .isApplicableTo(
+                      ByteRangeSpec.relativeLength(null, effectiveMax_0),
+                      ByteRangeSpec.explicit(null, effectiveMax_0),
+                      ByteRangeSpec.explicitClosed(null, effectiveMax_1)),
+              // expect that, we can read from 1 up to Long.MAX_VALUE - 1
+              RangeScenario.expectThat()
+                  .beginOffset(1L)
+                  .endOffset(effectiveMax_0)
+                  .endOffsetInclusive(effectiveMax_1)
```

Comment:
Is "inclusive" in terms of math or that the endOffset is part of the range e.g. `%d-%d`? If it's the former, I would think inclusive would be `effectiveMax_0` and not `effectiveMax_1`.

Reply:
The inclusive value would be the second arg in the header. So it's inclusive in terms of the index being included (`i <= endOffsetInclusive`), whereas endOffset would be excluded (`i < endOffset`).

Reply:
Converting into an example because my brain turned to mush; Am I understanding this correctly?

endOffset(N):
`Range: bytes=0-(N-1)` where N is exclusive it goes from (0-N-1)

endOffsetInclusive(N-1):
`Range: bytes=0-(N+1)` where N is inclusive it goes from (0-N)

Reply:
Imagine we have three ranges, A, B, C:
* A: `[0, N)` aka The set of all x such that 0 less than or equal to x and x less than N (`{x | 0 <= x < N}`)
* B: `[0, N]` aka The set of all x such that 0 less than or equal to x and x less than N (`{x | 0 <= x <= N}`)
* C: `0 + N` aka relative length

When converting A to an http range header, we know that the upper bound is exclusive, therefore we would produce `bytes=0-(N-1)`. For a range with an exclusive upper bound, `endOffset` will always equal `N` and `endOffsetInclusive` will equal `N-1`.

Contrasted by B, the upper bound of an inclusive range would produce `bytes=0-N`. Where `endOffset` will always equal `N+1` and `endOffsetInclusive` will equal `N`.

In the case of C, the upper bound is calculated relative to the beginning and would produce `bytes=0-(0+N-1)`. The `endOffset` will always be equal to `begin + length` whereas the `endOffsetInclusive` will be `begin + length - 1`.

This test is trying to say "Expect a range where the being offset is 1, the endOffset is effectiveMax_0, the endOffsetInclusive is effectiveMax_1, and the length is effectiveMax_1" and this should hold true for all three range declarations `[1 + effectiveMax_1]`, `[1, effectiveMax_0)` and `[0, effectiveMax_1]`.